File: Services/NoteMapper.cs
using OptionA.Tuner.Models;

namespace OptionA.Tuner.Services;

public static class NoteMapper
{
    /// <summary>
    /// Maps a detected frequency to the nearest note, computing cents offset
    /// </summary>
    /// <param name="frequency">Detected frequency in Hz</param>
    /// <param name="referenceA4">Reference pitch for A4 (e.g., 440.0)</param>
    /// <returns>NoteInfo with note name, octave, and cents offset</returns>
    public static NoteInfo Map(double frequency, double referenceA4 = 440.0)
    {
        // Calculate the number of semitones from A4
        var semitonesFromA4 = 12.0 * Math.Log2(frequency / referenceA4);

        // MIDI note number (A4 = 69)
        var midiNoteExact = 69.0 + semitonesFromA4;
        var midiNoteRounded = (int)Math.Round(midiNoteExact);

        // Cents offset from the nearest note
        var centsOffset = (midiNoteExact - midiNoteRounded) * 100.0;

        // Clamp cents to -50..+50
        centsOffset = Math.Clamp(centsOffset, -50.0, 50.0);

        // Note name and octave from MIDI note number
        // MIDI note 0 = C-1, so octave = (midiNote / 12) - 1
        var noteIndex = ((midiNoteRounded % 12) + 12) % 12;
        var octave = (midiNoteRounded / 12) - 1;

        // Target frequency for the nearest note
        var targetFrequency = referenceA4 * Math.Pow(2.0, (midiNoteRounded - 69) / 12.0);

        return new NoteInfo
        {
            Name = NoteInfo.NoteNames[noteIndex],
            Octave = octave,
            CentsOffset = centsOffset,
            TargetFrequency = targetFrequency,
            DetectedFrequency = frequency,
            MidiNote = midiNoteRounded
        };
    }
}

File: web/js/note-mapper.js
// OptionA Tuner — Note Mapper (12-TET)
// Ported from Services/NoteMapper.cs

const NOTE_NAMES = ['C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#', 'A', 'A#', 'B'];

/**
 * @typedef {Object} NoteInfo
 * @property {string} name        Note name (e.g. "A", "C#")
 * @property {number} octave      Octave number
 * @property {number} centsOffset Cents offset from perfect pitch (-50..+50)
 * @property {number} targetFreq  Exact frequency of the nearest 12-TET note
 * @property {number} detectedFreq Actual detected frequency
 * @property {number} midiNote    MIDI note number (A4 = 69)
 * @property {boolean} isInTune   |cents| <= 5
 * @property {boolean} isSharp    cents > 5
 * @property {boolean} isFlat     cents < -5
 */

/**
 * Map a detected frequency to the nearest 12-TET note.
 * @param {number} frequency    Detected frequency in Hz
 * @param {number} [referenceA4=440] Reference pitch for A4
 * @returns {NoteInfo}
 */
export function mapNote(frequency, referenceA4 = 440) {
    const semitonesFromA4 = 12.0 * Math.log2(frequency / referenceA4);
    const midiExact = 69.0 + semitonesFromA4;
    const midiRounded = Math.round(midiExact);

    let centsOffset = (midiExact - midiRounded) * 100.0;
    centsOffset = Math.max(-50, Math.min(50, centsOffset));

    const noteIndex = ((midiRounded % 12) + 12) % 12;
    const octave = Math.floor(midiRounded / 12) - 1;
    const targetFreq = referenceA4 * Math.pow(2, (midiRounded - 69) / 12);

    return {
        name: NOTE_NAMES[noteIndex],
        octave,
        centsOffset,
        targetFreq,
        detectedFreq: frequency,
        midiNote: midiRounded,
        isInTune: Math.abs(centsOffset) <= 5,
        isSharp: centsOffset > 5,
        isFlat: centsOffset < -5,
    };
}
